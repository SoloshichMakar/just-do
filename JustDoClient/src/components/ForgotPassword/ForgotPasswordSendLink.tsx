import React from "react";
import { Snackbar } from "@material-ui/core";
import JustDoLogo from "../../picture/JustDoLogo.png";
import JustDoComponent from "../../picture/JustDoComponent.png";
import BackArrow from "../../picture/ic_arrow_left.svg";
import FormControl from "@material-ui/core/FormControl";
import Input from "@material-ui/core/Input";
import { Link, useHistory } from "react-router-dom";
import { ColorButton } from "../style/CustomElements";
import { gql } from "apollo-boost";
import {
  emailHelperMessage,
  validateEmail,
} from "../../utils/validateUserDataFunction";
import { useMutation } from "@apollo/react-hooks";
import { Alert } from "@material-ui/lab";

interface IForgotPassword {
  email: string;
  message: string;
  setMessage: Function;
  errorMessage: Function;
  clearMessage: Function;
  textEmailChange: Function;
}

const SEND_INSTRUCTIONS = gql`
  mutation SendEmailPasswordRestore($email: String!) {
    sendEmailPasswordRestore(email: $email) {
      message
    }
  }
`;

let emailIsEdit = false;

const ForgotPassword: React.FC<IForgotPassword> = ({
  email,
  message,
  setMessage,
  errorMessage,
  clearMessage,
  textEmailChange,
}) => {
  const [sendEmailPasswordRestore] = useMutation(SEND_INSTRUCTIONS);
  const history = useHistory();

  function currentTextEmailChange(e: React.ChangeEvent<HTMLInputElement>) {
    emailIsEdit = true;
    const getTextAreaValue = e.target.value;
    textEmailChange(getTextAreaValue);
  }

  async function sendEmail() {
    try {
      if (!validateEmail(emailIsEdit, email) && emailIsEdit) {
        const response = await sendEmailPasswordRestore({
          variables: { email: email },
        });
        emailIsEdit = false;
        setMessage("Instruction to restore your password is sent to your email: " + email);
        textEmailChange("");
        history.push("/");
      }
    } catch (error) {
      errorMessage();
    }
  }

  const handleErrorClose = (event?: React.SyntheticEvent, reason?: string) => {
    if (reason === "clickaway") {
      return;
    }
    clearMessage();
  };

  return (
    <div id="forgot_password_main">
      <Snackbar
        open={message.length !== 0}
        autoHideDuration={10000}
        onClose={handleErrorClose}
      >
        <Alert onClose={handleErrorClose} severity="error">
          {message}
        </Alert>
      </Snackbar>
      <div id="forgot_password_main__logo_images">
        <img id="logo" src={JustDoLogo} />
        <img src={JustDoComponent} id="background_image" />
      </div>
      <div id="forgot_password_main__components">
        <Link id="arrow_link" to="/login">
          <img src={BackArrow} />
        </Link>
        <h1>Forgot Password</h1>
        <div id="components_helper_text">
          {" "}
          Please enter your email below to receive your <br /> password reset
          instruction{" "}
        </div>
        <div id="components__input_components">
          <div className="input_components__text_input">
            <FormControl id="textInput__email">
              <Input
                id="email_input"
                error={validateEmail(emailIsEdit, email)}
                type="text"
                value={email}
                onChange={currentTextEmailChange}
                placeholder="E-mail"
              />
              <div id="errorMessage">
                {emailHelperMessage(emailIsEdit, email)}
              </div>
            </FormControl>
          </div>

          <div id="input_components__color_button">
            <ColorButton
              id="color_button"
              variant="contained"
              color="primary"
              onClick={() => sendEmail()}
            >
              Send
            </ColorButton>
          </div>
        </div>
      </div>
    </div>
  );
};

export default ForgotPassword;
