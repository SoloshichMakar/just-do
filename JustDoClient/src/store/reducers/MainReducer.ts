import RegistrationReducer from "./RegistrationReducer";
import AuthorizationReducer from "./AuthorizationReducer";
import PasswordRestoreReducer from "./PasswordRestoreReducer";
import ForgotPasswordReducer from "./ForgotPasswordReducer";

import { combineReducers } from "redux";

const MainReducer = combineReducers({
  RegistrationReducer,
  AuthorizationReducer,
  PasswordRestoreReducer,
  ForgotPasswordReducer,
});

export default MainReducer;
