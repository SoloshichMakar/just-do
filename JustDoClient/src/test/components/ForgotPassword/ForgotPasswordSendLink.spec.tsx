import renderer from "react-test-renderer";
import { createStore } from "redux";
import MainReducer from "../../../store/reducers/MainReducer";
import { Provider } from "react-redux";
import React from "react";
import ForgotPassword from "../../../components/ForgotPassword/ForgotPasswordSendLink";
import ApolloClient from "apollo-boost";
import { ApolloProvider } from "react-apollo";
import { BrowserRouter } from "react-router-dom";

const props = {
  email: "",
  message: "",
  setMessage: Function,
  errorMessage: Function,
  clearMessage: Function,
  textEmailChange: Function,
};

const client = new ApolloClient({
  uri: ``,
});
const store = createStore(MainReducer);

describe("Forgot password component snapshot", () => {
  it("Forgot password component render correctly", () => {
    const tree = renderer
      .create(
        <ApolloProvider client={client}>
          <Provider store={store}>
            <BrowserRouter>
              <ForgotPassword
                email={props.email}
                clearMessage={props.clearMessage}
                errorMessage={props.errorMessage}
                message={props.message}
                setMessage={props.setMessage}
                textEmailChange={props.textEmailChange}
              />
            </BrowserRouter>
          </Provider>
        </ApolloProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
