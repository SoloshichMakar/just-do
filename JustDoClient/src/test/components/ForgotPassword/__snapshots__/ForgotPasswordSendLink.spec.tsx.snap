// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forgot password component snapshot Forgot password component render correctly 1`] = `
<div
  id="forgot_password_main"
>
  <div
    id="forgot_password_main__logo_images"
  >
    <img
      id="logo"
      src="JustDoLogo.png"
    />
    <img
      id="background_image"
      src="JustDoComponent.png"
    />
  </div>
  <div
    id="forgot_password_main__components"
  >
    <a
      href="/login"
      id="arrow_link"
      onClick={[Function]}
    >
      <img
        src="ic_arrow_left.svg"
      />
    </a>
    <h1>
      Forgot Password
    </h1>
    <div
      id="components_helper_text"
    >
       
      Please enter your email below to receive your 
      <br />
       password reset instruction
       
    </div>
    <div
      id="components__input_components"
    >
      <div
        className="input_components__text_input"
      >
        <div
          className="MuiFormControl-root"
          id="textInput__email"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="email_input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="E-mail"
              required={false}
              type="text"
              value=""
            />
          </div>
          <div
            id="errorMessage"
          />
        </div>
      </div>
      <div
        id="input_components__color_button"
      >
        <button
          className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-44 MuiButton-contained MuiButton-containedPrimary"
          disabled={false}
          id="color_button"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Send
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Registration snapshot Registration render correctly 1`] = `
<div
  id="forgot_password_main"
>
  <div
    id="forgot_password_main__logo_images"
  >
    <img
      id="logo"
      src="JustDoLogo.png"
    />
    <img
      id="background_image"
      src="JustDoComponent.png"
    />
  </div>
  <div
    id="forgot_password_main__components"
  >
    <a
      href="/login"
      id="arrow_link"
      onClick={[Function]}
    >
      <img
        src="ic_arrow_left.svg"
      />
    </a>
    <h1>
      Forgot Password
    </h1>
    <div
      id="components_helper_text"
    >
       
      Please enter your email below to receive your 
      <br />
       password reset instruction
       
    </div>
    <div
      id="components__input_components"
    >
      <div
        className="input_components__text_input"
      >
        <div
          className="MuiFormControl-root"
          id="textInput__email"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="email_input"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="E-mail"
              required={false}
              type="text"
              value=""
            />
          </div>
          <div
            id="errorMessage"
          />
        </div>
      </div>
      <div
        id="input_components__color_button"
      >
        <button
          className="MuiButtonBase-root MuiButton-root WithStyles(ForwardRef(Button))-root-44 MuiButton-contained MuiButton-containedPrimary"
          disabled={false}
          id="color_button"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Send
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
